src/builder/create_components.rs

```diff
@@ -14,26 +14,108 @@ pub enum CreateActionRow {
     InputText(CreateInputText),
 }
 
+#[derive(Serialize, Deserialize)]
+struct ActionRowJson {
+    #[serde(rename = "type")]
+    kind: u8,
+    components: Vec<serde_json::Value>,
+}
+
 impl serde::Serialize for CreateActionRow {
     fn serialize<S: serde::Serializer>(&self, serializer: S) -> Result<S::Ok, S::Error> {
         use serde::ser::Error as _;
 
-        serde_json::json!({
-            "type": 1,
-            "components": match self {
-                Self::Buttons(x) => serde_json::to_value(x).map_err(S::Error::custom)?,
-                Self::SelectMenu(x) => serde_json::to_value(vec![x]).map_err(S::Error::custom)?,
-                Self::InputText(x) => serde_json::to_value(vec![x]).map_err(S::Error::custom)?,
-            }
-        })
+        let components: Vec<serde_json::Value> = match self {
+            Self::Buttons(x) => {
+                let mut buttons = Vec::new();
+
+                // This needs to be a for loop so that we can use `?` to return
+                // early on error
```

Comment:
Don't know if these comments are helpful or not /shrug

Reply:
Btw, you may be interested in the "collect to Result" pattern in the future: `let buttons = x.iter().map(|button| serde_json::to_value(button).map_err(S::Error::custom)?).collect::<Result<Vec<_>, _>>()?;`

Reply:
Thanks for the recommendation! I'll have to check out how that works exactly, but I swapped it out for now 👍